alternator/executor.hh

```diff
@@ -70,6 +70,76 @@ public:
     std::string to_json() const override;
 };
 
+namespace parsed {
+class path;
+};
+
+// An attribute_path_map object is used to hold data for various attributes
+// paths (parsed::path) in a hierarchy of attribute paths. Each attribute path
+// has a root attribute, and then modified by member and index operators -
+// for example in "a.b[2].c" we have "a" as the root, then ".b" member, then
+// "[2]" index, and finally ".c" member.
+// Data can be added to an attribute_path_map using the add() function, but
+// requires that attributes with data not be *overlapping* or *conflicting*:
+//
+// 1. Two attribute paths which are identical or an ancestor of one another
+//    are considered *overlapping* and not allowed. If a.b.c has data,
+//    we can't add more data in a.b.c or any of its descendants like a.b.c.d.
+//
+// 2. Two attribute paths which need the same parent to have both a member and
+//    an index are considered *conflicting* and not allowed. E.g., if a.b has
+//    data, you can't add a[1]. The meaning of adding both would be that the
+//    attribute a is both a map and an array, which isn't sensible.
+//
+// These two requirements are common to the two places where Alternator uses
+// this abstraction to describe how a hierarchical item is to be transformed:
+//
+// 1. In ProjectExpression: for filtering from a full top-level attribute
+//    only the parts for which user asked in ProjectionExpression.
+//
+// 2. In UpdateExpression: for taking the previous value of a top-level
+//    attribute, and modifying it based on the instructions in the user
+//    wrote in UpdateExpression.
+
+template<typename T>
+class attribute_path_map_node {
+public:
+    using data_t = T;
+    // We need the extra shared_ptr<> here because libstdc++ unordered_map
+    // doesn't work with incomplete types :-( We couldn't use lw_shared_ptr<>
+    // because it doesn't work for incomplete types either. We couldn't use
+    // std::unique_ptr<> because it makes the entire object uncopyable. We
+    // don't often need to copy such a map, but we do have some code that
+    // copies an attrs_to_get object, and is hard to find and remove.
+    // The shared_ptr should never be null.
+    using members_t =  std::unordered_map<std::string, seastar::shared_ptr<attribute_path_map_node<T>>>;
+    // The indexes list is sorted because DynamoDB requires handling writes
+    // beyond the end of a list in index order.
+    using indexes_t = std::map<unsigned, seastar::shared_ptr<attribute_path_map_node<T>>>;
```

Comment:
Question: does DynamoDB allow arrays with holes - e.g. there's data at index 0, then nothing, and then data at index 99999999999? I'm just trying to understand why it's a map, and not a vector or deque - perhaps it was just easier to implement that way?

Reply:
Or can we perhaps have really long arrays, in which case it's better to split it and avoid continuous allocations?

Reply:
‎Maybe my comments are not clear enough, but in attribute_path_map we don't store the actual items - we store an (hierarchical) list of attribute paths (and values - hence a "map") - which are listed on ProjectionExpression or UpdateExpression. You're right that the item itself has an actual list, with no holes, but someone can pass `ProjectionExpression="a[1], a[7]" and that's the information we store in that map - that the user asked just for items 1 and 7.

I don't expect these maps to be long - they just contain a parsed version of the UpdateExpression or ProjectionExpression. But even if they are long, they shouldn't be contiguous - I don't think std::map is contiguous.
By the way, if you're curious why I used std::map and not std::unordered_map for the indexes, well, it's because of a corner case of what we have to do if the indexes in the UpdateExpression are beyond the end of the array, in which commands like a[1000]=2 or a[2000]=3 just mean add to the end of the array - but in this order (first the 1000 then the 2000).

Reply:
Ah, alright, makes sense